src/printer.js

```diff
@@ -1131,6 +1132,14 @@ function printExpression(path, options, print) {
           ])
         );
       }
+      const isAssociative = !!node.items[0].key;
+      const shouldBreak =
+        isAssociative &&
+        hasNewlineInRange(
+          options.originalText,
```

Comment:
I think I'd actually prefer the implementation I had before, because we can access the relevant part of the original source code via `node.loc.source`. I think `hasNewlineInRange` only exists because in JS the original source is only accessed "in full" via `options.originalText`.